src/ngraph/runtime/cpu/pass/cpu_fusion.cpp

```diff
@@ -60,6 +60,38 @@
 #include "ngraph/runtime/cpu/op/sigmoid_mul.hpp"
 #include "ngraph/util.hpp"
 
+// CPUFusion algorithm:
+// CPUFusion processes an input graph in an topological order(i.e. args before users)
+// Given the following graph:          Abs2
+//                                   /       \
+//                         Constant1         Add4 - Result5
+//                                   \      /
+//                                    Neg3
+//
+// The topological order would be : `Constant1`, `Abs2`, `Neg3`, `Add4`, `Result5`
+// Note, `Abs2` comes before `Neg3` as `Abs2`'s id = 2 is *less* than `Neg3`'s one (id = 3)
+// Next, CPUFusion will invoke matchers registered in an order registered in a c-tor
+// i.e. if a c-tor calls `construct_m1()`; `construct_m2()`; `construct_m3()`;
+// Matchers will be called as follows: `m1`, `m2`, `m3`
+// Matchers should only replace nodes in the graph that come before the current root
+// node in the topological order. For example, if Matcher matches Neg3, it should only
+// replace nodes `Abs2` and `Constant1` if needed
+// This gives Matchers a nice cascading property. For example, if m1 folds `Abs2(Constant1)`
+// and `m2` folds `Neg3(Constant1)` when `m3` is called on `Add4` it will discover that
+// both `Abs2` and `Neg3` were already replaced by constants, so `Add4` will also be folded into one.
+// If any Matcher succeeds the rest of the matchers will **not** be called.
+// E.g. if `m1` succeeds and replaces `Abs2` with a new constant, nor `m2` or `m3` will be called
+// However, sometimes, you will need more than one fusion occur on the same node.
+// In this case, you should be able to request another pass of CPUFusion.
+// To request another pass, you will need to register fusions in a callback:
+// i.e. you will need to pass `this` into a callback and then call `this->construct_X`
+// This will schedule another pass of CPUFusion with the following fusion.
+// This approach should only be used if you are either:
+// a) need more than one fusion occur on the same node
+// b) you are modifying nodes after the current node in the topological order
+// c) there's no linear order of fusions which will give
+//    the correct final fusion. i.e. the same fusion needs to occur before and after some other fusion
+
```

Comment:
shouldn't this comment be in graph_rewrite.hpp or somewhere non-cpu specific, may be in core_fusion.cpp?

Reply:
Agree.